Move Lazy Shape Inference functions to pytorch core

wconstab opened the pull request:
Test Plan: tested via lazy_tensor_staging branch (test_ptltc operator tests as well as torchbench inference)

Differential Revision: D34187999

bdhirsh commented on tools/codegen/api/lazy.py:
```diff
@@ -58,19 +63,36 @@ def process_ir_type(typ: Type) -> Union[BaseCType, VectorCType, OptionalCType, L
 
-def isValueType(typ: Union[Type, BaseCType, OptionalCType, ConstRefCType, MutRefCType,
-                           ListCType, ArrayRefCType, ArrayCType, VectorCType, TupleCType]) -> bool:
+def isValueType(typ: CType) -> bool:
```
Is this meant to fix https://github.com/pytorch/pytorch/issues/72852? (It looks like `isValueType()` never actually gets called with a normal `Type` if I'm following correctly, which makes this an easy fix)

wconstab replied:
yes, it is (to both)

bdhirsh commented on tools/codegen/api/lazy.py:
```diff
@@ -58,19 +63,36 @@ def process_ir_type(typ: Type) -> Union[BaseCType, VectorCType, OptionalCType, L
-        return typ.type == valueT
+        # I am regretting my naming conventions, but now we are wrapping at::scalar in
+        # lazy value, while preserving other 'scalar' types as scalars in the IR
+        return typ.type == valueT or typ.type == scalarT
```
I don't totally follow from the comment - can you give a code example?

It also doesn't seem like this really matches the description of the function: "This is equivalent to being Tensor-like", if we directly return true on scalar types.

wconstab replied:
lazy tensor tracing promotes at::Scalars to become lazy tensors.  This enables treating the scalars as inputs to the lazy trace and not burning them into the program.

it could use a bit of fresh paint.  i think i might want to refactor some of this stuff but probably in a separate PR.  I could use your suggestions for that too more generally.   but I don't think I want to make too many changes at this PR since it's aiming to just sync up master to what we are already using on staging.  Better to develop the new version on staging where we can test it directly.  (until we finally finish merging enough to master that we can build/run entirely on master's CI.)

bdhirsh commented on tools/codegen/api/lazy.py:
```diff
@@ -108,14 +132,15 @@ def __init__(self, func: FunctionSchema):
+            if curr_args is not None:
+                if isinstance(curr_args, TensorOptionsArguments):
+                    curr_args = curr_args.all()
+                keyword_arg_types.extend([NamedCType(arg.name, process_ir_type(arg.type)) for arg in curr_args])
```
Basic question - it looks like right now, `LazyIrSchema` has separate fields for positional and keyword arguments. Is that info used meaningfully anywhere?

If not, I think you could simplify a bunch of this logic into something like:
```
self.keyword_arg_types = [NamedCType(arg.name, process_ir_type(arg.type)) for ag in func.schema_order_arguments()]
```

That would help you avoid having to special-case `TensorOptionsArguments`, `SelfArgument`, and have to explicitly list out the `Arguments` field names (which I think would cause LTC codegen to break in pretty subtle ways if we changed the names of those fields in core codegen)

wconstab replied:
if you look at `lazy_ts_lowering.py` you'll see how we generate different code for kw args vs positional args.  I couldn't think of a cleaner way at the time but I agree it's a bit complex.

bdhirsh commented on tools/codegen/api/lazy.py:
```diff
@@ -108,14 +132,15 @@ def __init__(self, func: FunctionSchema):
         self.keyword_arg_types = tuple(keyword_arg_types)
         self.name = func.name
         self.returns = func.returns
+        self.wrapped_scalar_names = [arg.name for arg in func.schema_order_arguments() if isWrappedScalarType(arg.type)]
```
Ok, so the `LazyIrSchema` object knows "what are the C++ types and names of all arguments in my schema", and now it also knows "which of my arguments were originally scalars before processing", using this separate list.

I think the other "data-model-y" option you could take would be to represent that info per-argument using a new type (e.g. a `LTCArgument` that knows both its `NamedCType`, and whether it was a scalar argument). The advantage there is that your processing functions (`node_ctor_arg_rvalue_string`) can run on each argument in isolation, instead of having to take in the whole schema. But if this list is used in a really minimal/targeted way then maybe that's overkill.

wconstab replied:
i think this approach sounds good.  generally i think i want to separate out more 'refactoring' to a separate diff.  Lets collect ideas for that on a GH issue. i'll make one and link here: https://github.com/pytorch/pytorch/issues/72960

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -35,24 +42,55 @@ def node_ctor_arg_rvalue_string(arg: NamedCType) -> str:
+
+def aten_symbol(schema: LazyIrSchema) -> str:
+    missing_interned_strings = {
+        'sigmoid_backward',
```
side note: now that this is all in core, should we just add these symbols in core instead of having to handle them specially in codegen?

wconstab replied:
I actually added a handful of symbols to core, and was never sure why i had to do that, and ultimately figured this was a cleaner escape hatch.

if the idea is to have all the symbols be constants in core, should we just make a diff that adds all the symbols at once?  otherwise, it's not super clear why we have some and not others...

bdhirsh replied:
I was trying to look into where `at::aten::{symbol}` is actually defined - do you know where it is?

 eter Bell landed some changes recently that I think are meant to codegen all of the aten symbols, so I'm wondering if this probably was basically already solved for you. (I can see `sigmoid_backward` listed in the codegen'd file `build/aten/src/ATen/core/aten_interned_strings.h`)

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -246,19 +304,20 @@ class GenLazyShapeInferenceDefinition:
     tensor_class: str
 
     @method_with_native_function
+    # def gen_lazy_shape_inference_decl(f: NativeFunction, backend_index: BackendIndex, tensor_class: str) -> List[str]:
```
nit: commented code

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -246,19 +304,20 @@ class GenLazyShapeInferenceDefinition:
         # since we just use the meta function for structured kernels
         if not f.structured and f.structured_delegate is None:
-            shape_sig = ComputeShapeSignature(f)
+            shape_sig = ComputeShapeSignature(metadata.kernel, f)
```
This doesn't seem like a big deal - but any reason why we need to use the `metadata.kernel` to choose how to name the shape functions? For example, we could do something like `f.func.name.unambiguous_name()` instead

wconstab replied:
I can't answer off the top of my head.  I probably didn't know about unambiguous_name().  I think i needed metadata.kernel to overcome some ambiguity so this may be possible to substitute. I'll have to test it separately (back on staging) and then drop the update in if it works.

wconstab replied:
ok, so here is one case where they differ.

ipdb> p metadata.kernel
'arange_out'
ipdb> p f.func.name.unambiguous_name()
'arange_start_out'

I'm not sure what the right thing to do is.

bdhirsh replied:
leaving for a refactor sounds fine.

I think at a high level, what we probably want to have happen (based on what some of the other codegen in core does), is that the shape signatures are part of some API (the "lazy API", or "lazy shape API") that knows how to generate its name. That way at both the place where you define those functions in the codegen, and at the place where you call those functions, you can use the same api call to generate the name (and you don't have to remember whether or not it was actually using `metadata.kernel` or `f.unambiguous_name()`.

wconstab replied:
yea that makes sense. i'll reflect that in the other (refactor) issue.

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -174,37 +229,40 @@ def this_shape(i: int) -> str:
                 shapes_str = ','.join([this_shape(i) for i in range(returns_length)])
                 meta_out = "std::vector<Shape> shapes{" + shapes_str + "};"
 
+            # TODO: INTEGRATION POINT HERE:
```
what's the TODO for?

wconstab replied:
ah that should be removed..

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -174,37 +229,40 @@ def this_shape(i: int) -> str:
+        bridge_str = """auto result = torch::lazy::CreateAtenFromLtcTensor(
+                torch::lazy::LazyTensor::Create(std::move(node), *common_device));"""
 
-        assert len(value_types) > 0, f"Only supporting tensor ops so far, none found in {sig}"
```
did we lose the assert?

bdhirsh replied:
(NVM, i see the new one further up)

bdhirsh commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -174,37 +229,40 @@ def this_shape(i: int) -> str:
-    // TODO(alanwaketan): Quite a lot inefficient copy-by-value there. Let's optimize it.
-    {sig.decl(name=f"{self.class_method_name}::{schema.aten_name}")} {{
+    {sig.decl(name=f"{self.class_method_name}::{metadata.kernel}")} {{
+        {fallback_str}
```
Just confirming - so this PR also adds "per-op-defined eager fallback"?

wconstab replied:
the fallback thing here is just a debugging feature.  It lets you set an env variable that will force a particular operator to fall back.

bdhirsh commented on tools/codegen/gen_lazy_tensor.py:
```diff
@@ -144,7 +172,15 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
+    if shape_inference_hdr is not None:
+        expected_shape_infr_decls = list(
+            concat_map_codegen(
+                dest.GenLazyShapeInferenceDefinition(backend_indices[backend_key], tensor_class),
```
Do you mind writing out somewhere what the flow looks like for non-meta tensor shape inference? It'd help to have it listed somewhere:
(1) what all of the components are (the headers, shape inference kernels, their call-sites)
(2) which parts are codegen'd

bdhirsh commented on torch/csrc/lazy/core/shape_inference.cpp:
```diff
@@ -0,0 +1,540 @@
+
+std::vector<Shape> compute_shape_abs(const at::Tensor & self) {
+  if (self.is_complex()) {
+    const auto float_type = c10::toValueType(self.scalar_type());
```
oh, I think you might be hitting CI failures because `toValueType()` was recently changed to `toRealValueType()`

wconstab replied:
ahh, gotcha.  I will follow up on that. hopefully it is just a trivial rename otherwise i'll have to do some gymnastics to get toRealValueType on staging branch first and test it there..

wconstab commented on torch/csrc/lazy/core/shape_inference.cpp:
```diff
@@ -0,0 +1,540 @@
+ *
+ * READ THIS!
+ *
+ * 1. Beware: Tech Debt!
```
I could put a ref to read the 'tech debt' section in shape_inference.cpp in the codegen file.

I'm not sure what you mean about inplace codegen though?

bdhirsh reviewed (approved):
Approving, since the main purpose of this PR is to true up master with the staging branch and refactors are left as a followup item in https://github.com/pytorch/pytorch/issues/72960

wconstab committed:
Move Lazy Shape Inference functions to pytorch core (#72756)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/72756

Test Plan: tested via lazy_tensor_staging branch (test_ptltc operator tests as well as torchbench inference)

Reviewed By: bdhirsh

Differential Revision: D34187999

fbshipit-source-id: 0fb6856a111e76408f32de227f3e5dfa2d2eb28c
